Sources/Player/OfflineEngine/Internal/Storage/DBStorage/DBOfflineStorage.swift

```diff
@@ -0,0 +1,103 @@
+import Foundation
+import GRDB
+
+// MARK: - DBOfflineStorage
+
+class DBOfflineStorage {
+	private let dbQueue: DatabaseQueue
+
+	init(dbQueue: DatabaseQueue) {
+		self.dbQueue = dbQueue
+		try? initializeDatabase()
+	}
+}
+
+// MARK: OfflineStorage
+
+extension DBOfflineStorage: OfflineStorage {
+	// MARK: - Save OfflineEntry
+
+	func save(_ entry: OfflineEntry) throws {
+		let entryDTO = DBOfflineEntryDTO(from: entry)
+		try dbQueue.write { db in
+			try entryDTO.insert(db)
+		}
+	}
+
+	// MARK: - Get OfflineEntry by MediaProduct
+
+	func get(mediaProduct: MediaProduct) throws -> OfflineEntry? {
+		let entryDTO = try dbQueue.read { db in
+			try DBOfflineEntryDTO.filter(DBOfflineEntryDTO.Columns.productId == mediaProduct.productId).fetchOne(db)
+		}
+		return entryDTO?.toOfflineEntry()
+	}
+
+	// MARK: - Delete OfflineEntry by MediaProduct
+
+	func delete(mediaProduct: MediaProduct) throws {
+		_ = try dbQueue.write { db in
+			try DBOfflineEntryDTO.filter(DBOfflineEntryDTO.Columns.productId == mediaProduct.productId).deleteAll(db)
+		}
+	}
+
+	// MARK: - Update OfflineEntry
+
+	func update(_ entry: OfflineEntry) throws {
+		let entryDTO = DBOfflineEntryDTO(from: entry)
+		try dbQueue.write { db in
+			try entryDTO.update(db)
+		}
+	}
+
+	// MARK: - Get All CacheEntries
```

Comment:
[suggestion] I guess copy and paste issue
```suggestion
	// MARK: - Get All Offline Entries
```